state-sync/storage-service/server/src/lib.rs

```diff
@@ -0,0 +1,238 @@
+        ))
+    }
+
+    fn get_transactions_with_proof(
+        &self,
+        request: TransactionsWithProofRequest,
+    ) -> Result<StorageServiceResponse, Error> {
+        let transactions_with_proof = self.storage.get_transactions_with_proof(
+            request.proof_version,
+            request.start_version,
+            request.expected_num_transactions,
+            request.include_events,
+        )?;
+
+        Ok(StorageServiceResponse::TransactionsWithProof(
+            transactions_with_proof,
+        ))
+    }
+}
+
+/// The interface into local storage (e.g., the Diem DB) used by the storage
+/// server to handle client requests.
+pub trait StorageReaderInterface {
+    /// Returns a data summary of the underlying storage state.
+    fn get_data_summary(&self) -> Result<DataSummary, Error>;
+
+    /// Returns a list of transactions with a proof relative to the
+    /// `proof_version`. The transaction list is expected to contain *at most*
+    /// `expected_num_transactions` and start at `start_version`.
+    /// If `include_events` is true, events are also returned.
+    fn get_transactions_with_proof(
+        &self,
+        proof_version: u64,
+        start_version: u64,
+        expected_num_transactions: u64,
+        include_events: bool,
+    ) -> Result<TransactionListWithProof, Error>;
+
+    /// Returns a list of epoch ending ledger infos, starting at `start_epoch`
+    /// and ending *at most* at the `expected_end_epoch`.
+    fn get_epoch_ending_ledger_infos(
+        &self,
+        start_epoch: u64,
+        expected_end_epoch: u64,
+    ) -> Result<EpochChangeProof, Error>;
+
+    // TODO(joshlind): support me!
+    //
+    // Returns a list of transaction outputs with a proof relative to the
+    // `proof_version`. The transaction output list is expected to contain
+    // *at most* `expected_num_transaction_outputs` and start at `start_version`.
+    //fn get_transaction_outputs_with_proof(
+    //    &self,
+    //    proof_version: u64,
+    //    start_version: u64,
+    //    expected_num_transaction_outputs: u64,
+    //) -> Result<TransactionOutputListWithProof, Error>;
+
+    // TODO(joshlind): support me!
+    //
+    // Returns an AccountStateChunk holding a list of account states
+    // starting at the specified account key with *at most*
+    // `expected_num_account_states`.
+    //fn get_account_states_chunk(
+    //    version,
+    //    start_account_key,
+    //    expected_num_account_states: u64,
+    //) -> Result<AccountStateChunk, Error>
+
+    // TODO(joshlind): identify the exact API here to get the frozen subtrees!
+    //
+    // Returns the frozen subtrees at the specified version.
+    //fn get_frozen_subtrees(
+    //    version,
+    //) -> Result<Vec<HashValue>, Error>
+}
+
+/// The underlying implementation of the StorageReaderInterface, used by the
+/// storage server.
+pub struct StorageReader {
+    storage: Arc<RwLock<DbReaderWriter>>,
```

Comment:
Does it ever write to the DB? Why not a `Arc<dyn DbReader>`?

And why is the RwLock needed?

Reply:
@msmouse, it doesn't ever write to the db. However, I'm trying to be consistent in the state sync and storage service code. Everything will get an `Arc<RwLock<DbReaderWriter>>`. The reason for this is we'll have several state sync (and storage service threads) reading and writing to the db concurrently and it's simpler to give them all a shared RwLock object than to try and do some magic around reader and writer only. IIUC, the DiemDB struct today does not have a global lock, correct? If so, I want to avoid any scary situations.

Reply:
it looks like the `DbReaderWriter` is just

```rust
pub struct DbReaderWriter {
    pub reader: Arc<dyn DbReader>,
    pub writer: Arc<dyn DbWriter>,
}
```

so whoever is constructing the `StorageReader` just could just clone the inner `Arc<dyn DbReader>` and hand it down.

as far as concurrency, part of the `DbReader` API contract is to be thread safe, so you don't need any locks/synchronization to call it. after all, we're effectively doing read-only calls to an MVCC database here, which supports unlimited read concurrency

Reply:
What happens if a single DbReader call (e.g., [get_transactions()](https://github.com/diem/diem/blob/20d4d715c9216a892b85455fa64e1389713d4f67/storage/diemdb/src/lib.rs#L696) makes several independent calls to the database (e.g., get_transactions, get_transaction_info and get_events_by_version). Each of these calls is going to read different objects in the database. MVCC might guarantee that each read works, but there's no guarantees across several independent reads, correct? A writer might write between them and violate the higher level assumptions (at the DbReader API). Likewise with writes. Note: the `get_transactions()` example is a little simple (because you could argue it's bound by versions, which shouldn't be changed once written, but I'm not comfortable making that assumption for everything). Perhaps this has all been thought about before, though? Happy to be convinced, but I'm still not there yet 😄

Reply:
The DB interfaces always take a "ledger_version" parameter. If not a flaw, it never assumes the user asks for "the latest".

Reply:
@msmouse, yeah, my feeling is that in most cases we might be alright. But, race conditions can be very sneaky and I'm still not 100% sure we won't have issues. We're also moving to a world where state sync is going to be multi-threaded and writing to the DB concurrently (e.g., writing different epochs, account states, transactions in parallel, etc.). Thus, I want to do this simple thing now (i.e., wrap all uses of state sync and storage code in a single RwLock -- maybe even consensus to avoid potential races between state sync and consensus). If we find that the single RwLock becomes a point of contention, I'll happily optimize it, but I'd like to make sure that everything is safe and simple for now. Does that make sense?

Reply:
As long as we are on the same page.

Reply:
Absolutely, thanks @msmouse (and @phlip9)! I've noted this down as something to dig into in the future (once we have an initial end-to-end prototype). 😄

Reply:
But still, we shouldn't need to write the DB in concurrency. It's not worth it. 😜

Reply:
@msmouse.. I think we'll see. I just don't want to rule this out now, especially as our current design is going to be processing data chunks in parallel 😄